sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaThriftTable.java

```diff
@@ -74,55 +92,81 @@ private static Schema thriftSchema(
 
   @Override
   protected PTransform<PCollection<KV<byte[], byte[]>>, PCollection<Row>> getPTransformForInput() {
-    final @NonNull SchemaProvider schemaProvider = ThriftSchema.provider();
-    return new PTransform<PCollection<KV<byte[], byte[]>>, PCollection<Row>>() {
-      @Override
-      @SuppressWarnings("nullness")
-      public PCollection<Row> expand(PCollection<KV<byte[], byte[]>> input) {
-        return input
-            .apply(Values.create())
-            .apply(MapElements.into(typeDescriptor).via(BeamKafkaThriftTable.this::decode))
-            .setSchema(
-                schema,
-                typeDescriptor,
-                schemaProvider.toRowFunction(typeDescriptor),
-                schemaProvider.fromRowFunction(typeDescriptor))
-            .apply(Convert.toRows());
-      }
-    };
+    return new InputTransformer(typeDescriptor, coder, schema);
   }
 
-  private T decode(byte[] bytes) {
-    try {
-      return thriftCoder.decode(new ByteArrayInputStream(bytes));
-    } catch (IOException e) {
-      throw new IllegalStateException(e);
+  private static class InputTransformer<T extends TBase<?, ?>>
```

Comment:
What I meant was to create some SerializableFunction or SimpleFuntion outside of this package (e.g. in  io/thrift) that would do the encoding/decoding. Right now we'll have to implement encode/decode it in every table provider that supports thrift format. You wouldn't need InputTransformer and OutputTransformer then.

What I have in mind it's the same as is done here:
https://github.com/apache/beam/blob/1466db958ca22222ae8badf1f21a860267d46d31/sdks/java/extensions/protobuf/src/main/java/org/apache/beam/sdk/extensions/protobuf/ProtoMessageSchema.java#L132

And then here:
https://github.com/apache/beam/blob/1466db958ca22222ae8badf1f21a860267d46d31/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaProtoTable.java#L83

only for Thrift encoding/decoding

Reply:
Got it, makes sense. I tried to solve it locally for Kafka, as I think making this truly reusable/composable requires a larger refactoring that should follow up. But those functions do seem to rather belong in the `ThriftSchema` anyway, so I'll fix this.

Reply:
I had a closer look and it seems like the proto `bytesToRow` and `rowToBytes` code has no real reason to be proto specific, as it can work with any `SchemaProvider` and `Coder` pair. I'll try to extract that into `org.apache.beam.sdk.util` and reuse it from both proto and thrift (and perhaps others, if I find more like this). This way I'll also add support for shuffled schemas in the thrift implementation for free.

Reply:
Just pushed another fixup for this. It's pretty big and I'll try to draw your attention to the potentially dangerous/incompatible changes.

Reply:
I can revert the change breaking change in proto (i.e. `Class<?>` -> `Class<? extends Message>`) and use some ugly casts internally, if this is a concern.

Reply:
Nevermind, I just noticed the comments which read:

    // Other modules are not allowed to use non-vendored Message class

So I'll revert that change.